src/package.c

```diff
@@ -975,18 +976,72 @@ int apk_ipkg_add_script(struct apk_installed_package *ipkg,
 	return 0;
 }
 
-void apk_ipkg_run_script(struct apk_installed_package *ipkg,
-			 struct apk_database *db,
-			 unsigned int type, char **argv)
+static int run_script(char** argv, int root_fd, char* fn)
 {
+	char *script_name = basename(fn);
+	int status;
+	pid_t pid;
 	static char * const environment[] = {
 		"PATH=/usr/sbin:/usr/bin:/sbin:/bin",
 		NULL
 	};
+
+	pid = fork();
+	if (pid == -1) {
+		apk_error("fork() failed with error str: %s", apk_error_str(errno));
+		return -2;
+	}
+	if (pid == 0) {
+		umask(0022);
+		if (fchdir(root_fd) == 0 && chroot(".") == 0) {
+				execve(fn, argv, environment);
+				apk_error("execve() failed for script with error str: %s", apk_error_str(errno));
+		}
+		exit(1); /* should not get here */
+	}
+	waitpid(pid, &status, 0);
+	if (!WIFEXITED(status) || WEXITSTATUS(status) != 0) {
+		apk_error("%s: script exited with error %d", script_name, WEXITSTATUS(status));
+		return -1;
+	}
+	return 0;
+}
+
+int apk_run_hook_script(void *ctx, int dirfd, const char *file)
+{
+	char fn[PATH_MAX];
+	struct apk_database *db = (struct apk_database *)ctx;
+	int root_fd = db->root_fd;
+	HOOK_SCRIPT_TYPE stage = db->hook_script;
+	char *argv[] = { "hook-script", NULL, NULL, NULL };
+
+	if ((apk_flags & (APK_NO_SCRIPTS | APK_SIMULATE)) != 0)
+		return 0;
+
+	if (stage == PRE_HOOK)
+		snprintf(fn, sizeof(fn), "etc/apk/pre_script.d/" "%s", file);
```

Comment:
I would prefer here `"etc/apk/%s_script.d/%s"` and use stage as index to array with `{ "pre", "post" }`.